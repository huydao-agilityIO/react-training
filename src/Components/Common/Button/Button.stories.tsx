// Button.stories.ts|tsx

import type { Meta, StoryObj } from '@storybook/react';

import Button from './index';

const meta: Meta<typeof Button> = {
  title: 'Components/Common/Button',
  component: Button,
  parameters: {
    layout: 'centered'
  },
  tags: ['autodocs'],
  args: {
    type: 'button',
    sizes: 'sm',
    isDisabled: false,
    isLoading: false,
    onClick: () => alert('Active')
  },
  argTypes: {
    variant: {
      options: ['primary', 'secondary'],
      control: { type: 'inline-radio' }
    },
    sizes: {
      options: ['sm', 'md'],
      control: { type: 'inline-radio' }
    },
    type: {
      options: ['submit', 'button', 'reset'],
      control: { type: 'inline-radio' }
    }
  }
};

export default meta;
type Story = StoryObj<typeof Button>;

export const Primary: Story = {
  args: {
    type: 'submit',
    variant: 'primary',
    children: 'Primary'
  }
};

export const Secondary: Story = {
  args: {
    variant: 'secondary',
    children: 'Secondary'
  }
};

export const Disabled: Story = {
  args: {
    variant: 'primary',
    children: 'Disabled',
    isDisabled: true
  }
};

export const Loading: Story = {
  args: {
    variant: 'primary',
    children: 'Loading',
    isLoading: true
  }
};
