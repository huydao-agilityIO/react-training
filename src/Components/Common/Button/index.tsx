import { ReactNode, memo } from 'react';
import { ButtonProps, Button } from '@chakra-ui/react';

interface ButtonComponentProps extends ButtonProps {
  children: ReactNode;
  variants?: 'primary' | 'secondary';
  sizes?: 'sm' | 'md';
}

const ButtonComponent = ({
  children,
  variants = 'primary',
  sizes = 'sm',
  ...props
}: ButtonComponentProps) => {
  return (
    <Button variant={variants} size={sizes} {...props}>
      {children}
    </Button>
  );
};

export default memo(ButtonComponent);
