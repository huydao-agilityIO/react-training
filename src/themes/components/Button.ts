import { defineStyleConfig } from '@chakra-ui/react';

const primaryStyles = {
  bg: 'primary.300',
  borderColor: 'primary.300',
  color: 'light.300'
};

const secondaryStyles = {
  bg: 'dark.100',
  borderColor: 'dark.100',
  color: 'light.200'
};

const buttonTheme = defineStyleConfig({
  baseStyle: {
    borderWidth: 1,
    borderStyle: 'solid',
    fontWeight: 'medium',
    textTransform: 'capitalize'
  },
  variants: {
    primary: {
      ...primaryStyles,
      _hover: {
        bg: 'light.300',
        borderColor: 'secondary.500',
        color: 'primary.300',
        _disabled: {
          ...primaryStyles
        }
      }
    },
    secondary: {
      ...secondaryStyles,
      _hover: {
        bg: 'light.200',
        color: 'dark.100',
        _disabled: {
          ...secondaryStyles
        }
      }
    }
  },
  sizes: {
    sm: {
      fontSize: 'sm',
      lineHeight: '4',
      paddingX: 3,
      paddingY: 6
    },
    md: {
      fontSize: 'md',
      lineHeight: '5',
      paddingX: 5,
      paddingY: 8
    }
  },
  defaultProps: {
    size: 'sm',
    variant: 'primary'
  }
});

export default buttonTheme;
