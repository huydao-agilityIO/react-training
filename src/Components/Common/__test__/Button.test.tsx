import chakraWrapper from './utils';

// Components
import Button from '../Button';

describe('Button Component', () => {
  it('match snapshot with primary variant', () => {
    const { asFragment } = chakraWrapper.render(
      <Button variant="solid" size="md">
        Click me primary
      </Button>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('match snapshot with secondary variant', () => {
    const { asFragment } = chakraWrapper.render(
      <Button variant="secondary" size="md">
        Click me secondary
      </Button>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders button with disabled state', () => {
    chakraWrapper.render(<Button isDisabled>Click me</Button>);

    const button = chakraWrapper.screen.getByRole('button', {
      name: /click me/i
    });

    expect(button).toBeDisabled();
  });

  it('shows spinner if isLoading', () => {
    const { getByTestId } = chakraWrapper.render(
      <Button data-testid="btn" isLoading>
        Login
      </Button>
    );
    expect(getByTestId('btn')).toHaveAttribute('data-loading');

    expect(chakraWrapper.screen.getByText('Login')).not.toBeVisible();

    chakraWrapper.screen.getByText('Loading...');
  });

  it('triggers click event when clicked', () => {
    const mockOnClick = jest.fn();
    const { getByText } = chakraWrapper.render(
      <Button onClick={mockOnClick}>Click me</Button>
    );
    const btn = getByText(/click me/i);

    chakraWrapper.fireEvent.click(btn);

    expect(mockOnClick).toHaveBeenCalled();
  });
});
