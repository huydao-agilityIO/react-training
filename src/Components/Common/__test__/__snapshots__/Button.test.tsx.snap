// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component match snapshot with primary variant 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-19k5i17"
    type="button"
  >
    Click me primary
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`Button Component match snapshot with secondary variant 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-1l4yr35"
    type="button"
  >
    Click me secondary
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
